### PR TITLE
[OPIK-3195] [BE] Implement support for experiment tags with full CRUD

### PR DESCRIPTION
**Pull request by yaricom**
## Details
To support adding TAGS during evaluation on the SDK side we need to add TAGS support on the backend side.
We should implement complete CRUD operations support over experiment tags.

## Key changes

---

### 1. Database Layer

**File**: `src/main/resources/liquibase/db-app-analytics/migrations/000051_add_tags_to_experiments.sql`

- Added `tags` column to the `experiments` table
- Column type: `Array(String)` with default empty array
- Includes rollback script

```sql
ALTER TABLE ${ANALYTICS_DB_DATABASE_NAME}.experiments ON CLUSTER '{cluster}'
    ADD COLUMN IF NOT EXISTS tags Array(String) DEFAULT [];
```

---

#### Experiment.java
- Added `tags` field (`Set<String>`) to the experiment model
- Field is exposed in both Public and Write views

```java
@JsonView({Experiment.View.Public.class, Experiment.View.Write.class}) Set<String> tags,
```

#### ExperimentUpdate.java
- Added `tags` field for updating experiment tags
- Supports partial updates (null values are ignored)

```java
Set<String> tags,
```

---

### 3. DAO Layer

**File**: `src/main/java/com/comet/opik/domain/ExperimentDAO.java`

#### Create Operations
- Handles tag insertion during experiment creation
- Empty set is stored as empty array if tags are null

#### Read Operations
- Deserializes tags from database array to `Set<String>`
- Returns `null` if the tag set is empty (not empty set)

```java
.tags(Optional.ofNullable(row.get("tags", String[].class))
        .map(tags -> Arrays.stream(tags).collect(toSet()))
        .filter(set -> !set.isEmpty())
        .orElse(null))
```

#### Update Operations
- Supports updating tags through `ExperimentUpdate`
- Tags can be replaced, cleared, or left unchanged

#### SQL Queries
- All INSERT, SELECT, and UPDATE statements updated to include `tags` column

---

### 4. Test Coverage

**File**: `src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java`

Added 267 lines of comprehensive test coverage:

#### Creation Tests
- **Parameterized test** covering three scenarios:
  - Tags with values (`Set.of("tag1", "tag2", "tag3")`)
  - Empty tags (`Set.of()`)
  - Null tags (`null`)
- Uses `@ParameterizedTest` with `@MethodSource` for maintainability

#### Update Tests
- Replace tags with new values
- Clear tags (set to empty or null)
- Update tags multiple times
- Edge cases and validation

#### Test Refinement
- Refactored three duplicate test methods into a single parameterized test
- Reduced code duplication by 19 lines
- Improved test maintainability and readability

---

## Commit History

| Commit | Description | Lines Changed |
|--------|-------------|---------------|
| `2f04b8a1e` | Implemented CRUD operations for experiment TAGs | +37 |
| `f0ef3c2a5` | Add tests for handling experiment tags in creation and updates | +289 |
| `fed60616e` | Refactor and parameterize experiment tag tests | +20, -39 |
| `e10f08339` | Fixed formatting | +1, -2 |

---

### Tag Storage Strategy
- **Database**: `Array(String)` in ClickHouse
- **Java**: `Set<String>` for uniqueness and unordered semantics
- **API**: JSON array of strings

### Null Handling
- **Input**: `null` or empty set accepted during creation/update
- **Storage**: Both stored as empty array in database
- **Output**: Empty arrays returned as `null` (not empty set) for cleaner API responses

### Backward Compatibility
- Existing experiments without tags will have empty array by default
- Migration is non-breaking

---

## Change checklist
- [ ] User facing
- [ ] Documentation update

## Issues

- Resolves #
- OPIK-3741

## Testing

Added related unit tests.

## Documentation

No documentation changes

### CODE REVIEW

File: apps/opik-backend/src/main/java/com/comet/opik/api/Experiment.java [hunk 1]
```diff
@@ -29,6 +30,7 @@ public record Experiment(
         @JsonView({Experiment.View.Public.class, Experiment.View.Write.class}) String name,
         @Schema(implementation = JsonListString.class) @JsonView({Experiment.View.Public.class,
                 Experiment.View.Write.class}) JsonNode metadata,
+        @JsonView({Experiment.View.Public.class, Experiment.View.Write.class}) Set<String> tags,
         @JsonView({Experiment.View.Public.class, Experiment.View.Write.class}) ExperimentType type,
```
[the remark this answers is not in the document]
Comment (yaricom) [reply apps/opik-backend/src/main/java/com/comet/opik/api/Experiment.java:36]: not sure if it is relevant

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java [hunk 2]
```diff
@@ -5747,4 +5747,291 @@ void updateExperiment_whenMultiplePartialUpdates_thenLatestValuesPreserved() {
...
+                    .tags(Set.of())
+                    .build();
+
+            // when
+            var experimentId = experimentResourceClient.create(experiment, API_KEY, TEST_WORKSPACE);
+
+            // then - backend returns null for empty tags
+            var expectedExperiment = experiment.toBuilder().tags(null).build();
+            var retrievedExperiment = getAndAssert(experimentId, expectedExperiment, TEST_WORKSPACE, API_KEY);
+            assertThat(retrievedExperiment.tags()).isNullOrEmpty();
```
[the remark this answers is not in the document]
Comment (yaricom) [reply apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java]: Test combined

File: apps/opik-backend/src/main/resources/liquibase/db-app-analytics/migrations/000051_add_tags_to_experiments.sql [hunk 3]
```diff
@@ -0,0 +1,8 @@
+--liquibase formatted sql
+--changeset iaroslavomelianenko:000051_add_tags_to_experiments
+--comment: Add tags column to experiments table
+
+ALTER TABLE ${ANALYTICS_DB_DATABASE_NAME}.experiments ON CLUSTER '{cluster}'
+    ADD COLUMN IF NOT EXISTS tags Array(String) DEFAULT [];
```
Comment (BorisTkachenko) [inline apps/opik-backend/src/main/resources/liquibase/db-app-analytics/migrations/000051_add_tags_to_experiments.sql]: To be consistent with `tags` for other tables probably would be better to remove default value

File: apps/opik-backend/src/main/java/com/comet/opik/domain/ExperimentDAO.java [hunk 4]
```diff
@@ -771,6 +777,7 @@ INSERT INTO experiments (
                 <if(name)> :name <else> name <endif> as name,
                 workspace_id,
                 <if(metadata)> :metadata <else> metadata <endif> as metadata,
+                <if(tags)> :tags <else> tags <endif> as tags,
```
Comment (ldaugusto) [inline apps/opik-backend/src/main/java/com/comet/opik/domain/ExperimentDAO.java:780]: so we expect updates with a full new set of tags, not just the new ones right?

File: apps/opik-backend/src/main/java/com/comet/opik/domain/ExperimentDAO.java [hunk 5]
```diff
@@ -1417,6 +1434,10 @@ private ST buildUpdateTemplate(ExperimentUpdate experimentUpdate, String update)
             template.add("metadata", experimentUpdate.metadata().toString());
         }
 
+        if (experimentUpdate.tags() != null) {
```
Comment (ldaugusto) [inline apps/opik-backend/src/main/java/com/comet/opik/domain/ExperimentDAO.java:1439]: use CollectionUtils.notEmpty

File: apps/opik-backend/src/main/java/com/comet/opik/domain/ExperimentDAO.java [hunk 6]
```diff
@@ -1441,6 +1462,10 @@ private void bindUpdateParams(ExperimentUpdate experimentUpdate, Statement state
             statement.bind("metadata", experimentUpdate.metadata().toString());
         }
 
+        if (experimentUpdate.tags() != null) {
```
Comment (ldaugusto) [inline apps/opik-backend/src/main/java/com/comet/opik/domain/ExperimentDAO.java:1469]: use CollectionUtils.notEmpty

File: apps/opik-backend/src/main/resources/liquibase/db-app-analytics/migrations/000051_add_tags_to_experiments.sql [hunk 7]
```diff
@@ -0,0 +1,8 @@
...
+
+--rollback ALTER TABLE ${ANALYTICS_DB_DATABASE_NAME}.experiments ON CLUSTER '{cluster}' DROP COLUMN IF EXISTS tags;
```
Comment (ldaugusto) [inline apps/opik-backend/src/main/resources/liquibase/db-app-analytics/migrations/000051_add_tags_to_experiments.sql]: add EOL

File: apps/opik-backend/src/main/java/com/comet/opik/domain/ExperimentDAO.java [hunk 8]
```diff
@@ -921,6 +934,10 @@ private Publisher<Experiment> mapToDto(Result result) {
                     .projectId(RowUtils.getOptionalValue(row, "project_id", UUID.class))
                     .name(row.get("name", String.class))
                     .metadata(getJsonNodeOrDefault(row.get("metadata", String.class)))
+                    .tags(Optional.ofNullable(row.get("tags", String[].class))
+                            .map(tags -> Arrays.stream(tags).collect(toSet()))
+                            .filter(set -> !set.isEmpty())
```
Comment (ldaugusto) [inline apps/opik-backend/src/main/java/com/comet/opik/domain/ExperimentDAO.java]: CollectionUtils.notEmpty is safer, protects your from NPE

File: apps/opik-backend/src/main/java/com/comet/opik/api/Experiment.java [hunk 9]
```diff
@@ -29,6 +30,7 @@ public record Experiment(
         @JsonView({Experiment.View.Public.class, Experiment.View.Write.class}) String name,
         @Schema(implementation = JsonListString.class) @JsonView({Experiment.View.Public.class,
                 Experiment.View.Write.class}) JsonNode metadata,
+        @JsonView({Experiment.View.Public.class, Experiment.View.Write.class}) Set<String> tags,
```
Comment (ldaugusto) [inline apps/opik-backend/src/main/java/com/comet/opik/api/Experiment.java]: maybe adding some validation here?

@Valid 
@Size(max = 50, message = "Cannot have more than 50 tags")
Set<@NotBlank(message = "Tag must not be blank") @Size(max = 100, message = "Each tag must be at most 100 characters") String> tags

File: apps/opik-backend/src/main/java/com/comet/opik/domain/ExperimentDAO.java (hunk 4 above)
Comment (yaricom) [reply apps/opik-backend/src/main/java/com/comet/opik/domain/ExperimentDAO.java:780]: Yes, as far as I can see the `ExperimentDAO` is a bit unique on how it is updated. It is updated with INSERT rather than UPDATE as in `DatasetDAO` for example. So, I'm just following already existing logic.

File: apps/opik-backend/src/main/resources/liquibase/db-app-analytics/migrations/000051_add_tags_to_experiments.sql (hunk 7 above)
Comment (yaricom) [reply apps/opik-backend/src/main/resources/liquibase/db-app-analytics/migrations/000051_add_tags_to_experiments.sql]: Fixed

File: apps/opik-backend/src/main/java/com/comet/opik/domain/ExperimentDAO.java (hunk 5 above)
Comment (yaricom) [reply apps/opik-backend/src/main/java/com/comet/opik/domain/ExperimentDAO.java:1439]: Changed to Objects.nonNull(experimentUpdate.tags()) because it is expected to have empty but non null TAGs set when updating to have all tags removed.

File: apps/opik-backend/src/main/java/com/comet/opik/domain/ExperimentDAO.java (hunk 6 above)
Comment (yaricom) [reply apps/opik-backend/src/main/java/com/comet/opik/domain/ExperimentDAO.java:1469]: Changed to Objects.nonNull(experimentUpdate.tags()) because it is expected to have empty but non null TAGs set when updating to have all tags removed.

File: apps/opik-backend/src/main/java/com/comet/opik/domain/ExperimentDAO.java (hunk 8 above)
Comment (yaricom) [reply apps/opik-backend/src/main/java/com/comet/opik/domain/ExperimentDAO.java]: Fixed

File: apps/opik-backend/src/main/java/com/comet/opik/api/Experiment.java (hunk 9 above)
Comment (yaricom) [reply apps/opik-backend/src/main/java/com/comet/opik/api/Experiment.java]: Added validation

File: apps/opik-backend/src/main/java/com/comet/opik/domain/ExperimentDAO.java [hunk 10]
```diff
@@ -921,6 +935,10 @@ private Publisher<Experiment> mapToDto(Result result) {
                     .projectId(RowUtils.getOptionalValue(row, "project_id", UUID.class))
                     .name(row.get("name", String.class))
                     .metadata(getJsonNodeOrDefault(row.get("metadata", String.class)))
+                    .tags(Optional.ofNullable(row.get("tags", String[].class))
+                            .map(tags -> Arrays.stream(tags).collect(toSet()))
+                            .filter(CollectionUtils::isNotEmpty)
+                            .orElse(null))
```
[the remark this answers is not in the document]
Comment (yaricom) [reply apps/opik-backend/src/main/java/com/comet/opik/domain/ExperimentDAO.java:940]: Not relevant

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java [hunk 11]
```diff
@@ -5747,4 +5747,271 @@ void updateExperiment_whenMultiplePartialUpdates_thenLatestValuesPreserved() {
...
+            var expectedExperiment = expectedTags != null ? experiment : experiment.toBuilder().tags(null).build();
+            var retrievedExperiment = getAndAssert(experimentId, expectedExperiment, TEST_WORKSPACE, API_KEY);
+
+            if (expectedTags != null) {
+                assertThat(retrievedExperiment.tags())
+                        .isNotNull()
+                        .containsExactlyInAnyOrderElementsOf(expectedTags);
+            } else {
+                assertThat(retrievedExperiment.tags()).isNullOrEmpty();
+            }
```
Comment (thiagohora) [inline apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java]: Assert the whole object not just properties

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java [hunk 12]
```diff
@@ -5747,4 +5747,271 @@ void updateExperiment_whenMultiplePartialUpdates_thenLatestValuesPreserved() {
...
+                    .build();
+            experimentResourceClient.updateExperiment(experimentId, updateRequest, API_KEY, TEST_WORKSPACE,
+                    HttpStatus.SC_NO_CONTENT);
+
+            // then - verify tags are replaced
+            var expectedExperiment = experiment.toBuilder()
+                    .tags(updatedTags)
+                    .build();
+            var retrievedExperiment = getAndAssert(experimentId, expectedExperiment, TEST_WORKSPACE, API_KEY);
+            assertThat(retrievedExperiment.tags())
```
Comment (thiagohora) [inline apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java]: Same

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java [hunk 13]
```diff
@@ -5747,4 +5747,271 @@ void updateExperiment_whenMultiplePartialUpdates_thenLatestValuesPreserved() {
...
+                    .build();
+            experimentResourceClient.updateExperiment(experimentId, updateRequest, API_KEY, TEST_WORKSPACE,
+                    HttpStatus.SC_NO_CONTENT);
+
+            // then - verify tags are cleared (backend returns null for empty tags)
+            var expectedExperiment = experiment.toBuilder()
+                    .tags(null)
+                    .build();
+            var retrievedExperiment = getAndAssert(experimentId, expectedExperiment, TEST_WORKSPACE, API_KEY);
+            assertThat(retrievedExperiment.tags()).isNullOrEmpty();
```
Comment (thiagohora) [inline apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java]: Same

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java [hunk 14]
```diff
@@ -5747,4 +5747,271 @@ void updateExperiment_whenMultiplePartialUpdates_thenLatestValuesPreserved() {
...
+
+            // then - verify tags are preserved
+            var expectedExperiment = experiment.toBuilder()
+                    .name("New Name")
+                    .tags(initialTags)
+                    .build();
+            var retrievedExperiment = getAndAssert(experimentId, expectedExperiment, TEST_WORKSPACE, API_KEY);
+            assertThat(retrievedExperiment.tags())
+                    .isNotNull()
+                    .containsExactlyInAnyOrderElementsOf(initialTags);
```
Comment (thiagohora) [inline apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java]: Use `getAndAssert`

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java [hunk 15]
```diff
@@ -5747,4 +5747,271 @@ void updateExperiment_whenMultiplePartialUpdates_thenLatestValuesPreserved() {
...
+                    .type(ExperimentType.TRIAL)
+                    .status(ExperimentStatus.RUNNING)
+                    .build();
+            var retrievedExperiment = getAndAssert(experimentId, expectedExperiment, TEST_WORKSPACE, API_KEY);
+            assertThat(retrievedExperiment.tags())
+                    .isNotNull()
+                    .containsExactlyInAnyOrderElementsOf(updatedTags);
+            assertThat(retrievedExperiment.name()).isEqualTo("Updated Name");
+            assertThat(retrievedExperiment.type()).isEqualTo(ExperimentType.TRIAL);
+            assertThat(retrievedExperiment.status()).isEqualTo(ExperimentStatus.RUNNING);
```
Comment (thiagohora) [inline apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java]: Same

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java (hunk 11 above)
Comment (yaricom) [reply apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java]: @thiagohora Could you please provide an example of such test with full assertion. Thus, I would be able to follow.

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java [hunk 16]
```diff
@@ -5747,4 +5747,271 @@ void updateExperiment_whenMultiplePartialUpdates_thenLatestValuesPreserved() {
...
+            experimentResourceClient.updateExperiment(experimentId, secondUpdate, API_KEY, TEST_WORKSPACE,
+                    HttpStatus.SC_NO_CONTENT);
+
+            // then - verify second update (should replace first update tags)
+            var afterSecondUpdate = getAndAssert(experimentId, experiment.toBuilder().tags(secondUpdateTags).build(),
+                    TEST_WORKSPACE, API_KEY);
+            assertThat(afterSecondUpdate.tags())
+                    .containsExactlyInAnyOrderElementsOf(secondUpdateTags)
+                    .doesNotContainAnyElementsOf(firstUpdateTags)
+                    .doesNotContainAnyElementsOf(initialTags);
```
Comment (thiagohora) [inline apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java]: No need if you already did on `getAndAssert`

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java (hunk 11 above)
Comment (thiagohora) [reply apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java]: `getAndAssert` or `assertExperiments`

Comment (yaricom) [reply apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java]: As far as I understand the following line does assertion for all other properties

`var retrievedExperiment = getAndAssert(experimentId, expectedExperiment, TEST_WORKSPACE, API_KEY);`

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java (hunk 16 above)
Comment (yaricom) [reply apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java]: done

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java (hunk 15 above)
Comment (yaricom) [reply apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java]: Removed redundant `assertThat` calls validating experiment tags.

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java (hunk 14 above)
Comment (yaricom) [reply apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java]: Removed redundant `assertThat` calls validating experiment tags.

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java (hunk 13 above)
Comment (yaricom) [reply apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java]: Removed redundant `assertThat` calls validating experiment tags.

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java (hunk 12 above)
Comment (yaricom) [reply apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java]: Removed redundant `assertThat` calls validating experiment tags.

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java (hunk 11 above)
Comment (yaricom) [reply apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java]: Removed redundant `assertThat` calls validating experiment tags.

File: apps/opik-backend/src/main/java/com/comet/opik/api/ExperimentUpdate.java [hunk 17]
```diff
@@ -20,6 +21,7 @@
 public record ExperimentUpdate(
         @Pattern(regexp = NULL_OR_NOT_BLANK, message = "must not be blank") String name,
         JsonNode metadata,
+        Set<String> tags,
```
Comment (andrescrz) [inline apps/opik-backend/src/main/java/com/comet/opik/api/ExperimentUpdate.java]: The update request object needs the same validations as the one for creation (max number of tags, not blank tags, max tag chars etc.). Please align this.

File: apps/opik-backend/src/main/java/com/comet/opik/domain/ExperimentDAO.java [hunk 18]
```diff
@@ -821,6 +829,12 @@ private Publisher<? extends Result> insert(Experiment experiment, Connection con
                         .map(UUID::toString)
                         .orElse(""));
 
+        if (experiment.tags() != null) {
```
Comment (andrescrz) [inline apps/opik-backend/src/main/java/com/comet/opik/domain/ExperimentDAO.java]: Better checking that is not empty before performing the write, by using `CollectionUtils.isNotEmpty`.

File: apps/opik-backend/src/main/java/com/comet/opik/domain/ExperimentDAO.java [hunk 19]
```diff
@@ -1417,6 +1435,10 @@ private ST buildUpdateTemplate(ExperimentUpdate experimentUpdate, String update)
             template.add("metadata", experimentUpdate.metadata().toString());
         }
 
+        if (Objects.nonNull(experimentUpdate.tags())) {
```
Comment (andrescrz) [inline apps/opik-backend/src/main/java/com/comet/opik/domain/ExperimentDAO.java]: As commented above, all these writes below should be only performed when is not empty, with null safe check by `CollectionUtils.isNotEmpty`

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java [hunk 20]
```diff
@@ -5747,4 +5747,239 @@ void updateExperiment_whenMultiplePartialUpdates_thenLatestValuesPreserved() {
...
+            // given
+            var experiment = experimentResourceClient.createPartialExperiment()
+                    .tags(inputTags)
+                    .build();
+
+            // when
+            var experimentId = experimentResourceClient.create(experiment, API_KEY, TEST_WORKSPACE);
+
+            // then
+            var expectedExperiment = expectedTags != null ? experiment : experiment.toBuilder().tags(null).build();
```
Comment (andrescrz) [inline apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java]: Nit: this ternary conditional is kind of redundant. It's fine to always set tags to the `expectedTags`, even if null. Optionally, you could simplify this and merge inputTags and expectedTags to avoid duplication, and then refactor this conditional to set tags to null when inputTags/expectedTags is empty, otherwise you just set the expected ones.

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java [hunk 21]
```diff
@@ -5747,4 +5747,239 @@ void updateExperiment_whenMultiplePartialUpdates_thenLatestValuesPreserved() {
...
+            getAndAssert(experimentId, expectedExperiment, TEST_WORKSPACE, API_KEY);
+        }
+
+        private static java.util.stream.Stream<Arguments> createExperimentTagsProvider() {
+            return java.util.stream.Stream.of(
+                    Arguments.of(Set.of("tag1", "tag2", "tag3"), Set.of("tag1", "tag2", "tag3")),
```
Comment (andrescrz) [inline apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java]: Minor: better use random data to more endurable tests. You can leverage `PodamFactoryUtils.manufacturePojoSet` instead of hardcoded tag values.

This comment applies to all tests below with hardcoded tags.

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java [hunk 22]
```diff
@@ -5747,4 +5747,239 @@ void updateExperiment_whenMultiplePartialUpdates_thenLatestValuesPreserved() {
...
+                    Arguments.of(Set.of(), null),
+                    Arguments.of(null, null));
+        }
+
+        @Test
+        @DisplayName("when updating experiment tags, then tags are replaced correctly")
+        void updateExperimentTags_thenTagsReplacedCorrectly() {
```
Comment (andrescrz) [inline apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java:5783]: Minor: many these update tests can be better parameterised into a single one. See other examples in the code base. Anyhow, not a blocker.

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java (hunk 21 above)
Comment (andrescrz) [reply apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java]: Nit: please better use imports.

File: apps/opik-backend/src/main/java/com/comet/opik/domain/ExperimentDAO.java (hunk 19 above)
Comment (yaricom) [reply apps/opik-backend/src/main/java/com/comet/opik/domain/ExperimentDAO.java]: Here we need to check that it is not null. The empty set is OK - it is used to remove all tags.

File: apps/opik-backend/src/main/java/com/comet/opik/api/ExperimentUpdate.java (hunk 17 above)
Comment (yaricom) [reply apps/opik-backend/src/main/java/com/comet/opik/api/ExperimentUpdate.java]: Done

File: apps/opik-backend/src/main/java/com/comet/opik/domain/ExperimentDAO.java (hunk 18 above)
Comment (yaricom) [reply apps/opik-backend/src/main/java/com/comet/opik/domain/ExperimentDAO.java]: Done

File: apps/opik-backend/src/main/java/com/comet/opik/domain/ExperimentDAO.java (hunk 19 above)
Comment (yaricom) [reply apps/opik-backend/src/main/java/com/comet/opik/domain/ExperimentDAO.java]: The update operation can be used to remove all tags if empty set is provided.

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java (hunk 20 above)
Comment (yaricom) [reply apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java]: Removed  ternary conditional. The merge of `inputTags` and `expectedTags` will not work because it was requested to **remove DEFAULT value for tags in the DB**. Thus, if empty set is provided the `expectedTags` should be `null`.

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java (hunk 22 above)
Comment (yaricom) [reply apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java:5783]: Each update test has a bit different logic and test some important logic like removing tags if empty set provided. Not sure if combining them into one parameterized test improves readability and clarity.

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java (hunk 21 above)
Comment (yaricom) [reply apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java]: Fixed to use proposed `PodamFactoryUtils.manufacturePojoSet` where appropriate.

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java [hunk 23]
```diff
@@ -239,7 +239,7 @@ class ExperimentsResourceTest {
         APP = newTestDropwizardAppExtension(contextConfig);
     }
 
-    private final PodamFactory podamFactory = PodamFactoryUtils.newPodamFactory();
+    private static final PodamFactory podamFactory = PodamFactoryUtils.newPodamFactory();
```
Comment (andrescrz) [inline apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java]: This shouldn't be made static as unfortunately `PodamFactory` is not thread safe. Instead, you should just make `createExperimentTagsProvider` non-static. This works, as your test class is configured with per class lifecycle.

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java [hunk 24]
```diff
@@ -5747,4 +5747,240 @@ void updateExperiment_whenMultiplePartialUpdates_thenLatestValuesPreserved() {
...
+
+            // when
+            var experimentId = experimentResourceClient.create(experiment, API_KEY, TEST_WORKSPACE);
+
+            // then
+            var expectedExperiment = experiment.toBuilder().tags(expectedTags).build();
+            getAndAssert(experimentId, expectedExperiment, TEST_WORKSPACE, API_KEY);
+        }
+
+        private static java.util.stream.Stream<Arguments> createExperimentTagsProvider() {
```
Comment (andrescrz) [inline apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java]: Very nit: use imports here.

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java (hunk 22 above)
Comment (andrescrz) [reply apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java:5783]: These tests:
1. tags replaced correctly.
2. tags cleared correctly
3. existing tags preserved
4. tags preserved

Which always do the same steps:
1. Create experiment.
2. Update tags
3. assert expected experiment values, including tags.

You could merge them if you just parameterise the expected values (maybe the tags creation).

The rest of tests such as:
- Updating multiple fields.
- Experiments with multiple tags.
- Etc.

Make sense to be separated.

Anyhow, this is a minor suggestion from my end. Totally fine to keep this approach if you prefer. No need to bother to spend the time.

**Review by andrescrz — COMMENTED** [review]
Let's just revert the static from `podamFactory`, and this should be good to go.

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java (hunk 23 above)
Comment (yaricom) [reply apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java]: Done

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java (hunk 24 above)
Comment (yaricom) [reply apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ExperimentsResourceTest.java]: Removed

**Comment by yaricom** [comment]
@andrescrz I've already fixed podamFactory in this PR with latest commit.

**Review by andrescrz — APPROVED** [review]
